File: src/logic/ai.ts
/* @license Copyright (c) @kurehajime / source code: https://github.com/kurehajime/colamone_js */
'use strict';
import { Rule, MapArray, Hand, HandNode } from "./rule";
import { Eval } from "./eval";

export class Aijs {


  /** 
   * よく考える
   */
  static deepThinkAllAB(map: MapArray, turn_player: number, depth: number, a: number | undefined, b: number | undefined, nearwin: boolean, evalparam: number[][]): [Hand | undefined, number] {
    let best_score: number = turn_player * 9999999 * -1;
    let besthand;
    if (depth === 0) {
      best_score = Eval.evalMap(map, nearwin, evalparam);
      return [besthand, best_score];
    }
    if (a === void 0 || b === void 0) {
      a = 9999999 * turn_player * -1;
      b = 9999999 * turn_player;
    }

    let nodeList: HandNode[] = Rule.getNodeMap(map, turn_player);
    for (let i: number = 0; i < nodeList.length; i++) {
      let hand: Hand = nodeList[i][0];
      let map0 = nodeList[i][1];
      let sc: number = 0;
      // 必勝            
      let end: number = Rule.isEndX(map0, nearwin);
      if (end === turn_player) {
        return [hand, 999999 * turn_player];
      }
      // 必敗
      if (end === turn_player * -1) {
        if (besthand === void 0) {
          best_score = 999999 * turn_player * -1;
          besthand = hand;
        }
        continue;
      }
      if (Rule.isNoneNode(map0)) {
        sc = 0;
      } else {
        sc = Aijs.deepThinkAllAB(map0, turn_player * -1, depth - 1, b, a, nearwin, evalparam)[1];
      }
      if (besthand === void 0) {
        best_score = sc;
        besthand = hand;
      }
      if (turn_player === 1 && sc > best_score) {
        best_score = sc;
        besthand = hand;
      } else if (turn_player === -1 && sc < best_score) {
        best_score = sc;
        besthand = hand;
      }
      if (turn_player === 1 && a < best_score || turn_player === -1 && a > best_score) {
        a = best_score;
      }
      if (turn_player === 1 && b <= best_score || turn_player === -1 && b >= best_score) {
        break;
      }
    }
    return [besthand, best_score];
  }
  /** 
   * 考える
   */
  public static thinkAI(map: MapArray, turn_player: number, depth: number, a: number | undefined, b: number | undefined, evalparam: number[][] | undefined): [Hand | undefined, number] {
    let nearwin: boolean = false;
    let hand: [Hand | undefined, number] = [undefined, 0];
    let wkMap = new Int8Array(map);
    if (!evalparam) {
      evalparam = Eval.DEFAULT_EVALPARAM;
    }
    if (Rule.isEndX(wkMap, false) !== 0) {
      nearwin = true;
    }

    hand = Aijs.deepThinkAllAB(wkMap, turn_player, depth, a, b, nearwin, evalparam);
    if (hand[1]! * turn_player === -999999) {
      hand = Aijs.deepThinkAllAB(wkMap, turn_player, 1, a, b, nearwin, evalparam);
    }
    return hand;
  }
}

File: wasm/src/ai.rs
use crate::eval::{self, DEFAULT_EVALPARAM};
use crate::rule::{self, Hand, MapArray};

#[inline]
pub fn deep_think_all_ab(
    map: &MapArray,
    turn_player: isize,
    depth: isize,
    a: Option<isize>,
    b: Option<isize>,
    nearwin: bool,
    evalparam: [[isize; 6]; 9],
) -> (Option<Hand>, isize) {
    let mut best_score: isize = turn_player * 9999999 * -1;
    let mut besthand: Option<Hand> = None;
    let mut a = a;
    let mut b = b;
    if depth == 0 {
        best_score = eval::eval_map(map, nearwin, evalparam);
        return (None, best_score);
    }
    if a == None || b == None {
        a = Some(9999999 * turn_player * -1);
        b = Some(9999999 * turn_player);
    }

    let node_list = rule::get_node_map(map, turn_player);
    for node in node_list {
        let hand: Hand = node.0;
        let map0 = node.1;
        let sc: isize;
        // 必勝
        let end = rule::is_end_x(&map0, nearwin);
        if end == turn_player {
            return (Some(hand), 999999 * turn_player);
        }
        // 必敗
        if end == turn_player * -1 {
            if besthand == None {
                best_score = 999999 * turn_player * -1;
                besthand = Some(hand);
            }
            continue;
        }
        if rule::is_none_node(&map0) {
            sc = 0;
        } else {
            sc = deep_think_all_ab(&map0, turn_player * -1, depth - 1, b, a, nearwin, evalparam).1;
        }
        if besthand == None {
            best_score = sc;
            besthand = Some(hand);
        }
        if turn_player == 1 && sc > best_score {
            best_score = sc;
            besthand = Some(hand);
        } else if turn_player == -1 && sc < best_score {
            best_score = sc;
            besthand = Some(hand);
        }
        if turn_player == 1 && a.unwrap() < best_score
            || turn_player == -1 && a.unwrap() > best_score
        {
            a = Some(best_score);
        }
        if turn_player == 1 && b.unwrap() <= best_score
            || turn_player == -1 && b.unwrap() >= best_score
        {
            break;
        }
    }
    return (besthand, best_score);
}

pub fn think_ai(
    map: &MapArray,
    turn_player: isize,
    depth: isize,
    a: Option<isize>,
    b: Option<isize>,
    evalparam: Option<[[isize; 6]; 9]>,
) -> (Option<Hand>, isize) {
    let mut nearwin = false;
    let mut hand: (Option<Hand>, isize);
    let evalparam = evalparam.unwrap_or(DEFAULT_EVALPARAM);

    if rule::is_end_x(map, false) != 0 {
        nearwin = true;
    }
    hand = deep_think_all_ab(map, turn_player, depth, a, b, nearwin, evalparam);
    if hand.1 * turn_player == -999999 {
        hand = deep_think_all_ab(map, turn_player, 1, a, b, nearwin, evalparam);
    }
    return hand;
}
